File: app/src/main/java/com/shar2wy/moviesapp/models/trailerRepo/TrailerRemoteDataSource.java
package com.shar2wy.moviesapp.models.trailerRepo;

import android.content.Context;

import com.shar2wy.moviesapp.network.ApiClient;
import com.shar2wy.moviesapp.network.ApiService;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;

/**
 * Created by shar2wy
 * on 4/4/18.
 * Description: description goes here
 */
class TrailerRemoteDataSource {

    ApiService apiService;
    String apiKey = "38db4ee0b3b9fe63c95b9c835b73021f";

    public TrailerRemoteDataSource(Context context) {
        apiService = ApiClient.getClient(context).create(ApiService.class);
    }

    public Flowable<List<Trailer>> getTrailers(int id) {
        return apiService.getMovieTrailers(id, apiKey)
                .flatMap(trailersResponse -> Flowable.just(trailersResponse.getResults()))
                .flatMapIterable(trailers -> trailers)
                .map(trailer -> {
                    trailer.setMovieId(id);
                    return trailer;
                })
                .toList()
                .toFlowable()
                .subscribeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/java/com/shar2wy/moviesapp/models/trailerRepo/TrailerRemoteDataSource.kt
package com.shar2wy.moviesapp.models.trailerRepo

import android.content.Context
import com.shar2wy.moviesapp.network.ApiClient
import com.shar2wy.moviesapp.network.ApiService
import io.reactivex.Flowable

/**
 * Created by shar2wy
 * on 4/4/18.
 * Description: description goes here
 */
internal class TrailerRemoteDataSource(context: Context) {

    var apiService: ApiService = ApiClient.getInstance(context).getService(ApiService::class.java)
    var apiKey = "38db4ee0b3b9fe63c95b9c835b73021f"

    fun getTrailers(id: Int): Flowable<List<Trailer>> {
        return apiService.getMovieTrailers(id, apiKey)
                .flatMap { trailersResponse -> Flowable.just(trailersResponse.results) }
                .flatMapIterable { trailers -> trailers }
                .map { trailer ->
                    trailer.movieId = id
                    trailer
                }
                .toList()
                .toFlowable()
    }
}
